// allure-js-commons/dist/esm/src/current/AllureGroup.js
import { AllureTest } from "./AllureTest.js";
import { ExecutableItemWrapper } from "./ExecutableItemWrapper.js";
import { fixtureResult, testResultContainer } from "./constructors.js";
export class AllureGroup {
    runtime;
    testResultContainer = testResultContainer();
    constructor(runtime) {
        this.runtime = runtime;
    }
    startGroup(name) {
        const group = new AllureGroup(this.runtime);
        this.testResultContainer.children.push(group.uuid);
        group.name = name || "Unnamed";
        return group;
    }
    startTest(name, start) {
        const test = new AllureTest(this.runtime, start);
        this.testResultContainer.children.push(test.uuid);
        test.name = name || "Unnamed";
        return test;
    }
    endGroup() {
        this.runtime.writeGroup(this.testResultContainer);
    }
    get uuid() {
        return this.testResultContainer.uuid;
    }
    get name() {
        return this.testResultContainer.name || "";
    }
    set name(name) {
        this.testResultContainer.name = name;
    }
    addBefore() {
        const result = fixtureResult();
        this.testResultContainer.befores.push(result);
        return new ExecutableItemWrapper(result);
    }
    addAfter() {
        const result = fixtureResult();
        this.testResultContainer.afters.push(result);
        return new ExecutableItemWrapper(result);
    }
}
